src/integer_mod_q/mat_polynomial_ring_zq/arithmetic/mul.rs

```diff
@@ -0,0 +1,215 @@
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Panics if the dimensions of `self` and `other` do not match for multiplication.
+    /// - Panics if the moduli mismatch
+    fn mul(self, other: Self) -> Self::Output {
+        self.mul_safe(other).unwrap()
+    }
+}
+
+arithmetic_trait_borrowed_to_owned!(
+    Mul,
+    mul,
+    MatPolynomialRingZq,
+    MatPolynomialRingZq,
+    MatPolynomialRingZq
+);
+arithmetic_trait_mixed_borrowed_owned!(
+    Mul,
+    mul,
+    MatPolynomialRingZq,
+    MatPolynomialRingZq,
+    MatPolynomialRingZq
+);
+
+impl MatPolynomialRingZq {
+    /// Implements multiplication for two [`MatPolynomialRingZq`] values.
+    ///
+    /// Parameters:
+    /// - `other`: specifies the value to multiply with `self`
+    ///
+    /// Returns the product of `self` and `other` as a [`MatPolynomialRingZq`].
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let a = MatZq::from_str("[[2,1],[1,2]] mod 7").unwrap();
+    /// let b = MatZq::from_str("[[1,0],[0,1]] mod 7").unwrap();
+    ///
+    /// let c: MatZq = a.mul_safe(&b).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type
+    /// [`MathError::MismatchingMatrixDimension`] if the dimensions of `self`
+    ///  and `other` do not match for multiplication.
+    /// - Returns a [`MathError`] of type
+    /// [`MathError::MismatchingModulus`] if the moduli mismatch.
+    pub fn mul_safe(&self, other: &Self) -> Result<Self, MathError> {
+        if self.modulus != other.modulus {
+            return Err(MathError::MismatchingModulus(format!(
+                " Tried to add matrixes with moduli '{}' and '{}'.",
+                self.get_mod(),
+                other.get_mod()
+            )));
+        }
+        if self.get_num_columns() != other.get_num_rows() {
+            return Err(MathError::MismatchingMatrixDimension(format!(
+                "Tried to multiply a '{}x{}' matrix and a '{}x{}' matrix.",
+                self.get_num_rows(),
+                self.get_num_columns(),
+                other.get_num_rows(),
+                other.get_num_columns()
+            )));
+        }
+
+        let mut new =
+            MatPolynomialRingZq::new(self.get_num_rows(), other.get_num_columns(), &self.modulus)
+                .unwrap();
+        unsafe {
+            fmpz_poly_mat_mul(
+                &mut new.matrix.matrix,
+                &self.matrix.matrix,
+                &other.matrix.matrix,
+            );
+        }
+        new.reduce();
+
+        Ok(new)
+    }
+}
+
+#[cfg(test)]
+mod test_mul {
+    use super::MatPolynomialRingZq;
+    use crate::{integer::MatPolyOverZ, integer_mod_q::ModulusPolynomialRingZq};
+    use std::str::FromStr;
+
+    const BITPRIME64: u64 = u64::MAX - 58;
+
+    /// Checks if matrix multiplication works fine for squared matrices
```

Comment:
```suggestion
    /// Checks if matrix multiplication works fine for squared matrices.
```
Also applies to the other comments